### PR TITLE
WIP: parallel execution without Makefile - demo

### PR DESCRIPTION
**Pull request by themperek**
A very unpolished demo for #696 that removes the need for Makefiles (for discussion).
Libraries are build using setuptools Extension.
Allow basic discovery by pytest (better integration TBD).
Better portability.

The use case would be to run `pytest -n 16` which would discover all the test and run them in parallel/distributed using [pytest-xdist](https://github.com/pytest-dev/pytest-xdist).

Usage is to add somewhere in the test module (no Makefile needed):
```python3
def test_run():
    Run(sources=['../hdl/dff.v'], toplevel='dff')
```
and run `pytest`

Comments, suggestions help highly appreciated.

### CODE REVIEW

File: cocotb/run.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        "-s",
+        toplevel,
+        "-g2012",
+    ] + sources_abs
+    print(" ".join(comp_cmd))
+    process = Popen(comp_cmd).wait()
+
+    cmd = ["vvp", "-M", libs_dir, "-m", "gpivpi", "sim_build/sim.vvp"]
+    print(" ".join(cmd))
+    process = Popen(cmd).wait()
```
Comment (eric-wieser) [inline cocotb/run.py]: `subprocess.check_call(cmd)` would be much more idiomatic here and everywhere else.

File: cocotb/run.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        os.path.join(os.path.abspath(lib_path), "gpivpi.vpl"),
+    )
+    _symlink_force(
+        os.path.join(os.path.abspath(lib_path), "libvpi.so"),
+        os.path.join(os.path.abspath(lib_path), "cocotb.vpl"),
+    )
+    _symlink_force(
+        os.path.join(os.path.abspath(lib_path), "libsim.so"),
+        os.path.join(os.path.abspath(lib_path), "simulator.so"),
+    )
```
Comment (eric-wieser) [inline cocotb/run.py]: Don't think will work on windows, sadly

File: cocotb/run.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    dist = Distribution()
+    dist.parse_config_files()
+
+    libcocotbutils = Extension(
+        "libcocotbutils",
+        include_dirs=[share_dir + "/include"],
+        sources=[share_dir + "/lib/utils/cocotb_utils.c"],
+    )
+
+    lib_path = _build_lib(libcocotbutils, dist)
```
Comment (eric-wieser) [inline cocotb/run.py]: What happens if you set `dist.ext_modules = [libcocotbutils, libgpilog, libcocotb, libgpi]` and build them all at once?

File: cocotb/clock.py [hunk 4]
```diff
@@ -30,10 +30,11 @@
 import os
 import itertools
 
-if "SPHINX_BUILD" in os.environ:
-    simulator = None
-else:
+if "COCOTB_SIM" in os.environ:
     import simulator
+else:
+    simulator = None
```
Comment (eric-wieser) [inline cocotb/clock.py]: Might be worth putting this in its own PR and merging it separately - looks obviously worthwhile.

Comment (themperek) [reply cocotb/clock.py]: Right!

File: cocotb/run.py (hunk 2 above)
Comment (themperek) [reply cocotb/run.py]: Right. There will be more things not working on Windows ;-(

File: cocotb/run.py (hunk 1 above)
Comment (themperek) [reply cocotb/run.py]: Done

File: cocotb/run.py (hunk 3 above)
Comment (themperek) [reply cocotb/run.py]: I do not know [yet\ how to st/set build directory so I take it after libcocotbutils one is compiled ;-(

File: cocotb/run.py [hunk 5]
```diff
@@ -151,25 +157,28 @@ def Run(sources, toplevel, module):
     my_env["COCOTB_SIM"] = "1"
     my_env["MODULE"] = module
 
-    os.makedirs("sim_build", exist_ok=True)
+    sim_build_dir = os.path.join(run_dir_name, "sim_build")
+    os.makedirs(sim_build_dir, exist_ok=True)
+    sim_comopile_file = os.path.join(sim_build_dir, "sim.vvp")
```
Comment (eric-wieser) [inline cocotb/run.py]: typo: compile?

File: cocotb/clock.py (hunk 4 above)
Comment (imphil) [reply cocotb/clock.py]: Yeah, seems this was forgotten when doing the conversion.

File: cocotb/run.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+        toplevel,
+        "-g2012",
+    ] + sources_abs
+    print(comp_cmd)
+    print(" ".join(comp_cmd))
+    process = subprocess.check_call(comp_cmd)
+
+    cmd = ["vvp", "-M", libs_dir, "-m", "gpivpi", sim_comopile_file]
+    print(" ".join(cmd))
+    process = subprocess.check_call(cmd)
```
Comment (imphil) [inline cocotb/run.py]: That's the tricky part: having a way to do something similar for all simulators. You could have a look at edalize, a python library (used in fusesoc) to solve exactly this problem.

Comment (themperek) [reply cocotb/run.py]: Right, will check. Makefiles are not better ;-)

Comment (themperek) [reply cocotb/run.py]: If I am not mistaken edalize generates a Makefile and then runs make.  Something I would like to avoid if possible?. 

Something else would be to add cocotb integration to edalize so you can use the same definition for synthesis etc. (I guess only useful for FPGAs).

Comment (imphil) [reply cocotb/run.py]: The approach edalize takes depends a bit on the tool: sometimes it uses Makefiles as intermediate way to call the tool, sometimes it doesn't. Actually the current thinking is to get rid of the Makefiles again as they make things harder on Windows. But at the same time, Windows development moves in a direction that makes the use of Unix tools easier, so there are probably good reasons to go either way.


Beyond that, integrating cocotb with fusesoc has been on my todo list for ages, and that's definitely something I'll pick up again soon.

File: cocotb/result.py [hunk 7]
```diff
@@ -77,6 +77,8 @@ def __init__(self, retval):
 
 
 class TestComplete(Exception):
+    __test__ = False
```
Comment (eric-wieser) [inline cocotb/result.py]: Not sure this is the right approach - I've found it's better to point pytest at your test directory rather than mark things all over your code as "not a test". Perhaps I'm missing something though.

File: cocotb/run.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+def _symlink_force(target, link_name):
+    try:
+        os.symlink(target, link_name)
+    except OSError as e:
+        if e.errno == errno.EEXIST:
+            os.remove(link_name)
+            os.symlink(target, link_name)
+        else:
+            raise e
```
Comment (eric-wieser) [inline cocotb/run.py]: Better without the `e`, the stack trace will be shorter

File: cocotb/run.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+
+    my_env = os.environ
+    my_env["LD_LIBRARY_PATH"] = libs_dir + ":" + sysconfig.get_config_var("LIBDIR")
+
+    python_path = ":".join(sys.path)
+    my_env["PYTHONPATH"] = python_path + ":" + libs_dir
+    my_env["TOPLEVEL"] = toplevel
+    my_env["TOPLEVEL_LANG"] = "verilog"
+    my_env["COCOTB_SIM"] = "1"
+    my_env["MODULE"] = module
```
Comment (eric-wieser) [inline cocotb/run.py]: A neat trick would be to pass the test functions between embedded and non-embedded python using pickle, like how `multiprocessing` does.

File: cocotb/result.py (hunk 7 above)
Comment (themperek) [reply cocotb/result.py]: By default, it will look at this since name start with `Test` and it will cause a working. One way is to rename it other to leave the warning.

File: cocotb/run.py (hunk 9 above)
Comment (themperek) [reply cocotb/run.py]: Can be done but I guess we keep Makefile compatibility for now?

File: cocotb/clock.py (hunk 4 above)
Comment (garmin-mjames) [reply cocotb/clock.py]: Is the simulator even needed in this file?

Comment (themperek) [reply cocotb/clock.py]: Seems not ;-)

**Comment by themperek** [comment]
Work will continue as an external extension for now: https://github.com/themperek/cocotb-test
